### PR TITLE
[BitSail-437][Connector] V1 interface supports reading data from MongoDB source

### PR DESCRIPTION
**Pull request by healchow**
### Prepare a Pull Request

- Title: [BitSail-437][Connector] V1 interface supports reading data from MongoDB source

- Fixes: #437

> **Please change the PR title and the related issue number refer to the following description.**
>
> The above *XYZ* must be replaced with the actual [GitHub Issue](https://github.com/bytedance/bitsail/issues) number, indicating that the current PR is used to solve the problem described in this Issue.
>
> If you don't find an Issue that matches your changes, please [Create Issue](https://github.com/bytedance/bitsail/issues/new/choose) first, then commit this PR.
>
> For more info about the contributing guide, see: [Contributing Guide](https://github.com/bytedance/bitsail/blob/master/website/en/community/contribute.md)

### Motivation

V1 interface supports reading data from MongoDB source.

### Modifications

Refer the #104, migrate MongoDB legacy connector to adapt the V1 interface.

### Verifying this change

**Please pick either of the following options.**

- [x] This change added tests and can be verified as follows:

  - Added integration tests for MongoDB source: `MongoDBSourceITCase`

### Documentation

- Does this pull request introduce a new feature? (yes)
- If yes, how is the feature documented? (not applicable)
- If a feature is not documented yet in this PR, please create a follow-up issue for adding the documentation (provide later...)

### CODE REVIEW

File: bitsail-connectors/connector-mongodb/src/main/java/com/bytedance/bitsail/connector/mongodb/source/MongoDBSource.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import com.bytedance.bitsail.connector.mongodb.source.split.MongoDBSourceSplit;
+import com.bytedance.bitsail.connector.mongodb.source.split.MongoDBSourceSplitCoordinator;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Source of MongoDB
+ */
+public class MongoDBSource implements Source<Row, MongoDBSourceSplit, EmptyState>, ParallelismComputable {
```
[the remark this answers is not in the document]
Comment (healchow) [reply bitsail-connectors/connector-mongodb/src/main/java/com/bytedance/bitsail/connector/mongodb/source/MongoDBSource.java:42]: Sorry, I don't quite understand what you mean, where is the class name of the integration test wrong?

File: bitsail-connectors/connector-mongodb/src/main/java/com/bytedance/bitsail/connector/mongodb/source/reader/MongoDBRowDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+import java.util.Base64;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Deserializer for row data of MongoDB.
+ */
+public class MongoDBRowDeserializer {
```
[the remark this answers is not in the document]
Comment (healchow) [reply bitsail-connectors/connector-mongodb/src/main/java/com/bytedance/bitsail/connector/mongodb/source/reader/MongoDBRowDeserializer.java:47]: In theory, it is possible. But now there are very few usage examples, I am not very familiar with this architecture, and it is a bit difficult to change.

In addition, this PR is mainly to migrate the legacy interface to the V1 version, and other further optimizations, and then solve them one by one, is it OK?

File: bitsail-connectors/connector-mongodb/src/main/java/com/bytedance/bitsail/connector/mongodb/source/MongoDBSource.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (healchow) [reply bitsail-connectors/connector-mongodb/src/main/java/com/bytedance/bitsail/connector/mongodb/source/MongoDBSource.java:42]: I moved the MongoDB integration test to bitsail-test-integration module, please have a look, thanks.

**Comment by healchow** [comment]
The integration test failed, but I saw that this was caused by the failure to download the jar of Kudu.

And it did not execute the test cases I submitted this time. How to deal with this situation?


image

**Review by BlockLiu — APPROVED** [review]

File: bitsail-connectors/connector-mongodb/src/main/java/com/bytedance/bitsail/connector/mongodb/source/split/MongoDBSourceSplitCoordinator.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+      int readerIndex = ReaderSelector.getReaderIndex(readerNum);
+      splitAssignmentPlan.computeIfAbsent(readerIndex, k -> new HashSet<>()).add(split);
+      LOG.info("Will assign split {} to the {}-th reader", split.uniqSplitId(), readerIndex);
+    }
+  }
+
+  @Override
+  public void addReader(int subtaskId) {
+    LOG.info("Found reader {}", subtaskId);
+    tryAssignSplitsToReader();
```
[the remark this answers is not in the document]
Comment (healchow) [reply bitsail-connectors/connector-mongodb/src/main/java/com/bytedance/bitsail/connector/mongodb/source/split/MongoDBSourceSplitCoordinator.java:80]: In the tryAssignSplitsToReader method, there was already removed the task registered before re-assign.

**Pull request merged**